First version of microservice

pepeArenas opened the pull request:
This project contains all the loigic related to insert and retrive products from DB

Commit:
First version of microservice

Commit:
First version of web application

jahv commented on Management/.mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1 @@
+distributionUrl=https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip
```
Not needed file

jahv commented on Management/mvnw:
```diff
@@ -0,0 +1,225 @@
+#!/bin/sh
```
Not needed file

pepeArenas replied:
done

jahv commented on Management/mvnw.cmd:
```diff
@@ -0,0 +1,143 @@
+@REM ----------------------------------------------------------------------------
```
Not needed file

jahv commented on Management/pom.xml:
```diff
@@ -0,0 +1,106 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Change the 2 maven projects as maven modules so you'll have one single maven project with 2 modules

pepeArenas replied:
done

jahv commented on Management/src/main/java/com/stack/jspdemo/controller/ErrorController.java:
```diff
@@ -0,0 +1,21 @@
+public class ErrorController {
+
+	@ExceptionHandler(RuntimeException.class)
+	public ModelAndView handleException(HttpServletRequest request, HttpServletResponse response, Exception ex) {
```
I think it's fine to have a single class to handle exceptions, however, you should give more information in order to give support. Let's imagine the application throws the exception, how do you know the root cause?

jahv commented on Management/src/main/resources/application.properties:
```diff
@@ -0,0 +1,4 @@
+spring.mvc.view.prefix=/WEB-INF/view/
```
`/WEB-INF/view/` path doesn't exist

pepeArenas replied:
done

jahv commented on Management/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,52 @@
+
+	@RequestMapping(value = "/product", method = RequestMethod.POST)
+	public String showAddedProductPage(ModelMap model, @Valid Product product, BindingResult result) {
+		if (product.getName().equals("jesus")) {
```
I understand this is to validate the correct functionality of your error handler, but, is not better to have a test case for that and leave the main flow clean?

jahv commented on Management/src/main/java/com/stack/jspdemo/controller/ProductREST.java:
```diff
@@ -0,0 +1,24 @@
+package com.stack.jspdemo.controller;
```
What is this for? I don't see any use of this endpoint

pepeArenas replied:
Done

jahv commented on Management/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,64 @@
+	@Size(min = 10, message = "Enter at least 10 characters")
+	private String name;
+	private String model;
+	private String price;
```
Why is `price` a String?

pepeArenas replied:
done

jahv commented on Management/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	public Product(Integer id, String name, String model, String price) {
+		super();
```
Why do you call the `super()`?

pepeArenas replied:
done

jahv commented on Management/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	public Product(String name) {
+		super();
```
Why do you call the `super()`?
Try to keep your code organized, constructors at the very start of the class,  then getters/setters, and so on. Like here, you have setters, then another constructor then more setters

pepeArenas replied:
done

jahv commented on Management/src/main/java/com/stack/jspdemo/service/ProductService.java:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	public Product addProduct(Product product) {
+		if (product.getId() == null) {
```
I assume this class is a mock for your saving/getting the data.
My concern is the business logic on this, according to you, if the product has a null id, then you assign a new one based on the counter, let's imagine we have concurrent requests, they all have the same id, and theoretically, id should be unique

pepeArenas replied:
done

jahv commented on Management/src/main/resources/application.properties:
```diff
@@ -0,0 +1,4 @@
+spring.mvc.view.prefix=/WEB-INF/view/
+spring.mvc.view.suffix=.jsp
+logging.level.org.org.springframework.web=DEBUG
```
Why are you interested in this information?

pepeArenas replied:
Deleted, just for development purpouse

jahv commented on Management/src/main/webapp/WEB-INF/jsp/menu.jsp:
```diff
@@ -0,0 +1,7 @@
+
+<%@include file="commons/header.jspf"%>
+<%@include file="commons/navigation.jspf"%>
+<div class="container">
```
What is the reason of this empty `div`?

pepeArenas replied:
This is a part of a template, so I just left this one and remove this line from the JSP fragments where use this template, I don`t know, is this the better approach?

jahv commented on Management/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,64 @@
+@Component
+public class Product {
+	private Integer id;
+	@Size(min = 10, message = "Enter at least 10 characters")
```
Why 10 chars as the minimum?

pepeArenas replied:
Change to 45 to align with my DB definition

pepeArenas replied:
done

jahv commented on Management/src/main/webapp/WEB-INF/jsp/product.jsp:
```diff
@@ -0,0 +1,21 @@
+				<form:label path="name">Price:</form:label>
+			<form:input path="price" type="text" class="form-control"
+				required="required" />
+			<form:errors path="name" cssClass="text-warning"></form:errors>
```
Is not better to add the error under the `name` field?

pepeArenas replied:
done

jahv commented on Management/src/main/webapp/WEB-INF/jsp/productAdded.jsp:
```diff
@@ -0,0 +1,5 @@
+<html>
```
When is this one used?

pepeArenas replied:
deleted

jahv commented on ProductApp/.gitignore:
```diff
@@ -0,0 +1,29 @@
+/target/
```
You just need one `.gitignore` file, place it in the root folder

jahv commented on ProductApp/pom.xml:
```diff
@@ -0,0 +1,82 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Add this as a maven module

pepeArenas replied:
done

jahv commented on ProductApp/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,39 @@
+package com.stack.jspdemo.controller;
```
Why do you have 2 similar controllers, this one and the one from Management?

pepeArenas replied:
I will removed when I introduce the rabbitmq that miss

pepeArenas replied:
done

jahv commented on ProductApp/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,39 @@
+	@PostMapping("/products")
+	public ResponseEntity<Object> addProduct(@Valid @RequestBody Product product) {
+		Product savedProduct = service.save(product);
+		URI location = ServletUriComponentsBuilder.fromCurrentRequest().path("/{id}")
```
What is this for?

pepeArenas replied:
This was the case that you explain me yesterday, the idea is remove rest controller  expose a listener in order to receive the messages and then call the appropriate method, but the way I have been working I have duplicate functionality that will be moved from the common library in next steps

pepeArenas replied:
done

jahv commented on ProductApp/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,127 @@
+	@Size(min = 3, max = 45, message = "Name should be at leats 3 characters")
+	private String name;
+	private String model;
+	private String price;
```
Why price is String?

pepeArenas replied:
done

jahv commented on ProductApp/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,127 @@
+package com.stack.jspdemo.model;
```
Why do you have another Product class

pepeArenas replied:
This will be moved from the common library when I upload the common library

pepeArenas replied:
done

jahv commented on ProductApp/src/main/java/com/stack/jspdemo/service/ProductService.java:
```diff
@@ -0,0 +1,32 @@
+public class ProductService {
+	public static List<Product> products = new ArrayList<>();
+
+	private static int userCount = 3;
```
What is this for?

pepeArenas replied:
done

jahv commented on ProductApp/src/main/resources/application.properties:
```diff
@@ -0,0 +1,17 @@
+spring.application.name=ProductApp
+server.port=8000
+
+spring.datasource.url=jdbc:mysql://localhost:3306/ManagementProduct
```
We cannot test this app since we don't have the DB and the credentials, please upload the docker files for having this container

pepeArenas replied:
done

raoolgr commented on Management/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,52 @@
+@Controller
+public class ProductController {
+
+	@Autowired
```
Is it necessary to use `@Autowired` annotation here?

pepeArenas replied:
Its necessary, otherwise it will throws an null pointer exception or what do you refer about?

raoolgr replied:
If we use constructor injection it is not necessary to use `@Autowired` as Spring will automatically inject those beans. 
For more information you can check this link: https://docs.spring.io/spring-boot/docs/current/reference/html/using-boot-spring-beans-and-dependency-injection.html

pepeArenas replied:
done

raoolgr commented on Management/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,52 @@
+	ProductService productService;
+	@Autowired
+	Product product;
+
```
What about using constructor injection?

pepeArenas replied:
Removed this it was unnesesary

raoolgr commented on Management/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,64 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
Why are you making a value class a spring component?

pepeArenas replied:
It was unnecesary mark as a Component

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,39 @@
+@RestController
+public class ProductController {
+
+	@Autowired
```
Same here about using `@Autowired` annotation and constructor injection

pepeArenas replied:
Delete this file because it was not necessary anymore

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/controller/ProductController.java:
```diff
@@ -0,0 +1,39 @@
+		return service.findAll();
+	}
+
+	@PostMapping("/products")
```
Shall we specify what kind of message this endpoint consumes?

pepeArenas replied:
Yes, Antonio explain me yesterday this part is incorrect and here I will set up a listener which will receive rest, process it and persist/list the payload, It will be ready when upload the rabbitmq part

pepeArenas replied:
done

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,127 @@
+
+	@Override
+	public String toString() {
+		return "Product [id=" + id + ", name=" + name + ", model=" + model + ", price=" + price + "]";
```
Why not using `String.format()` instead of concatenating the string like this?

pepeArenas replied:
done

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	@Override
+	public int hashCode() {
```
In this moment is it really necessary to override `hashCode` and `equals()` ?

pepeArenas replied:
Removed

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/repository/ProductRepository.java:
```diff
@@ -0,0 +1,9 @@
+
+import com.stack.jspdemo.model.Product;
+
+public interface ProductRepository extends CrudRepository<Product, Long>, ProductRepositoryCustom{
```
Why not implementing instead `JpaRepository<E, R>` ? 
This interface already provides us a `findAll()` method, so it wouldn't be necessary to implement a `getAllProducts()` method

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/repository/ProductRepositoryCustom.java:
```diff
@@ -0,0 +1,12 @@
+public interface ProductRepositoryCustom {
+	List<Product> getAllProducts();
+
+	void insertProduct(Product product);
```
why not using `save()` method? this one is already provided by `CrudRepository` and `JpaRepostory`

raoolgr commented on Management/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,64 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class Product {
```
Why not making `Product` a `final` class?

pepeArenas replied:
Yes of course but I really don't know what would be the technical reason for that, can explain me why?

raoolgr commented on Management/src/main/webapp/WEB-INF/jsp/commons/footer.jspf:
```diff
@@ -0,0 +1,8 @@
+<script
+		src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
```
Please make this code more legible

pepeArenas replied:
done

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/model/Product.java:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
Please check item 10 of Effective Java. Most of this logic can be reduced using `instanceof`.
Also, please avoid using the IDE's autogenerated code.

pepeArenas replied:
removed

raoolgr commented on ProductApp/src/main/java/com/stack/jspdemo/repository/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,30 @@
+
+	@Override
+	public void insertProduct(Product product) {
+		System.out.println(product.getName());
```
Use a Logger instead of `System.out.println()`

pepeArenas replied:
done

Commit:
Remove unnecesary files

wombstrat commented on Management/src/main/java/com/stack/jspdemo/controller/ProductREST.java:
```diff
@@ -0,0 +1,24 @@
+
+	@Autowired
+	ProductService productService;
+	@Autowired
```
Aren't you injecting a singleton instance?
Consider using constructor based injections, as @ragore09 said comments above.

pepeArenas replied:
done

wombstrat commented on Management/src/main/java/com/stack/jspdemo/controller/WelcomeController.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.web.bind.annotation.RequestMethod;
+
+@Controller
+public class WelcomeController {
```
Is it needed to use public here?

pepeArenas replied:
Done but I have a question, is this because I do not need access this class outside the package?

wombstrat commented on Management/src/main/java/com/stack/jspdemo/service/ProductService.java:
```diff
@@ -0,0 +1,36 @@
+import com.stack.jspdemo.model.Product;
+
+@Service
+public class ProductService {
```
can't you use interfaces here?

wombstrat commented on Management/src/test/java/com/stack/jspdemo/ManagementApplicationTests.java:
```diff
@@ -0,0 +1,16 @@
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class ManagementApplicationTests {
```
If your tests aren't ready to be updated avoid pushing empty tests.

pepeArenas replied:
done

wombstrat commented on ProductApp/pom.xml:
```diff
@@ -0,0 +1,82 @@
+		<dependency>
+			<groupId>mysql</groupId>
+			<artifactId>mysql-connector-java</artifactId>
+			<version>8.0.12</version>
```
use properties here

wombstrat commented on ProductApp/pom.xml:
```diff
@@ -0,0 +1,82 @@
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-devtools</artifactId>
+			<scope>runtime</scope>
```
Nitpick questions: why runtime?

pepeArenas replied:
Well devtools is use in order to redeploy with any change during development phase so I feel this is the reason why is runtime needed

wombstrat commented on ProductApp/src/main/java/com/stack/jspdemo/service/ProductService.java:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	public Product save(Product product) {
+		// products.add(product);
```
remove comments like this.

pepeArenas replied:
done

wombstrat commented on ProductApp/src/main/resources/application.properties:
```diff
@@ -0,0 +1,17 @@
+spring.application.name=ProductApp
```
Prefer YML files

Commit:
Remove log level, remove nested div container and correct path to read jsp's

Commit:
Delete ProductRest.java, productAdded.jsp, remove logic from controller that throws exception as a test, remove super from constructor and remove logic from service that generate an id

Commit:
Remove hash code from Product in ProductApp project
Remove autowired when instanciate Product in ProductController,
Remove counter from productService

Commit:
Organize constructors and remove super() call
Change name to min 3 and max 45 characters
Remove component annotation from product pojo
Remove hash code and equals

Commit:
Format footer, header, navigation
Remove sout
Remove comment line

Commit:
Se refactoriza a proyecto por modulos en lugar de proyectos independientes

Commit:
Se agrega parte de rabbit mq

Commit:
Se agregan configuracines docker para mysql

Commit:
Primera version de commons library

Commit:
Configuracion de docker

Commit:
Pruebas unitarias,
Observaciones de respuesta cuando un servicio no esta arriba, correccion de varias observaciones

Commit:
Se corrige path de prueba unitaria

Commit:
Pruebas unitarias

Commit:
Se remueven autowired de variable de instancia y se ponen en contructor

jahv commented:
Closed
